### ISSUE TITLE
`WITH DOCKER` has 10 second delay

### ISSUE THREAD

**Issue by rrjjvv** [earthly/earthly#4066]
### What went wrong?

First, this is not an Earthly bug, and secondly, it's more of a regression or parity issue.

TL;DR; the issue is with the introduction of OTEL in dockerd `25.x`, which is present in `earthly/dind:alpine` images (but not present in `earthly/dind`).

I had noticed for a while that builds using `WITH DOCKER` seemed to have a lag at the end of that step, but that was usually a minor annoyance in the context of large builds (and had assumed it was associated with my use of many large images).  While iterating through #4063, where only a busybox image was pulled (and not even run), it was clear that lag was artificial.  Counting 'out loud', it seemed to always be around 10 seconds, which sounded like a hardcoded timeout (and it is).

```Earthfile
VERSION 0.7

stock-ubuntu:
    FROM earthly/dind
    WITH DOCKER
        RUN date && docker version && echo
    END

stock-alpine:
    FROM earthly/dind:alpine
    WITH DOCKER
        RUN date && docker version && echo
    END

better-alpine:
    FROM earthly/dind:alpine
    ENV OTEL_TRACES_EXPORTER=none
    WITH DOCKER
        RUN date && docker version && echo
    END
```

I added some diagnostics to the wrapper script:
```
diff --git a/buildkitd/dockerd-wrapper.sh b/buildkitd/dockerd-wrapper.sh
index 69acd36c..ee5e1d40 100755
--- a/buildkitd/dockerd-wrapper.sh
+++ b/buildkitd/dockerd-wrapper.sh
@@ -112,7 +112,10 @@ execute() {
     if [ "$EARTHLY_START_COMPOSE" = "true" ]; then
         docker_compose_cmd down --remove-orphans
     fi
+    echo "$(date) going to stop dockerd"
     stop_dockerd
+    echo "$(date) done stopping dockerd"
+    print_dockerd_logs | tail
     return "$exit_code"
 }
 ```

Older docker daemon (pre-OTEL):
```
$ time earthly-dev -P --no-cache +stock-ubuntu 
 <snip>
       +stock-ubuntu | Mon Apr 29 16:32:01 UTC 2024
<snip>
       +stock-ubuntu | Server: Docker Engine - Community
       +stock-ubuntu |  Engine:
       +stock-ubuntu |   Version:          20.10.14
<snip>
       +stock-ubuntu | Mon Apr 29 16:32:01 UTC 2024 going to stop dockerd
       +stock-ubuntu | Mon Apr 29 16:32:03 UTC 2024 done stopping dockerd
       +stock-ubuntu | time="2024-04-29T16:32:00.897644991Z" level=info msg="Loading containers: done."
       +stock-ubuntu | time="2024-04-29T16:32:00.908810712Z" level=info msg="Docker daemon" commit=87a90dc graphdriver(s)=overlay2 version=20.10.14
       +stock-ubuntu | time="2024-04-29T16:32:00.908890557Z" level=info msg="Daemon has completed initialization"
       +stock-ubuntu | time="2024-04-29T16:32:00.943077992Z" level=info msg="API listen on /var/run/docker.sock"
       +stock-ubuntu | time="2024-04-29T16:32:01.744972262Z" level=info msg="Processing signal 'terminated'"
       +stock-ubuntu | time="2024-04-29T16:32:01.745548365Z" level=info msg="stopping event stream following graceful shutdown" error="<nil>" module=libcontainerd namespace=moby
       +stock-ubuntu | time="2024-04-29T16:32:01.745837346Z" level=info msg="Daemon shutdown complete"
       +stock-ubuntu | time="2024-04-29T16:32:01.745840242Z" level=info msg="stopping healthcheck following graceful shutdown" module=libcontainerd
       +stock-ubuntu | time="2024-04-29T16:32:01.745858858Z" level=info msg="stopping event stream following graceful shutdown" error="context canceled" module=libcontainerd namespace=plugins.moby
       +stock-ubuntu | ==== End dockerd logs ====
              output | --> exporting outputs
<snip>
real	0m4.869s
user	0m0.248s
sys	0m0.170s
```

Newer daemon:
```
$ time earthly-dev -P --no-cache +stock-alpine
 <snip>
       +stock-alpine | Mon Apr 29 16:34:33 UTC 2024
<snip>
       +stock-alpine | Server:
       +stock-alpine |  Engine:
       +stock-alpine |   Version:          25.0.2
<snip>
       +stock-alpine | Mon Apr 29 16:34:33 UTC 2024 going to stop dockerd
       +stock-alpine | Mon Apr 29 16:34:44 UTC 2024 done stopping dockerd
       +stock-alpine | time="2024-04-29T16:34:33.487348365Z" level=info msg="Docker daemon" commit=fce6e0ca9bc000888de3daa157af14fa41fcd0ff containerd-snapshotter=false storage-driver=overlay2 version=25.0.2
       +stock-alpine | time="2024-04-29T16:34:33.487431240Z" level=info msg="Daemon has completed initialization"
       +stock-alpine | time="2024-04-29T16:34:33.527681699Z" level=info msg="API listen on /var/run/docker.sock"
       +stock-alpine | time="2024-04-29T16:34:33.546067804Z" level=info msg="Processing signal 'terminated'"
       +stock-alpine | time="2024-04-29T16:34:33.546971694Z" level=info msg="stopping event stream following graceful shutdown" error="<nil>" module=libcontainerd namespace=moby
       +stock-alpine | 2024/04/29 16:34:43 traces export: context deadline exceeded: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing: dial tcp: address /dev/otel-grpc.sock: missing port in address"
       +stock-alpine | time="2024-04-29T16:34:43.549047768Z" level=info msg="Daemon shutdown complete"
       +stock-alpine | time="2024-04-29T16:34:43.549112582Z" level=info msg="stopping healthcheck following graceful shutdown" module=libcontainerd
       +stock-alpine | time="2024-04-29T16:34:43.549132468Z" level=info msg="stopping event stream following graceful shutdown" error="context canceled" module=libcontainerd namespace=plugins.moby
       +stock-alpine | ==== End dockerd logs ====
              output | --> exporting outputs
<snip>
real	0m14.395s
user	0m0.286s
sys	0m0.148s
```

As you can see, it takes ten seconds longer, and it's clear that it's a timeout given for traces to be exported, and that is indeed the default:

> Timeout: Maximum time the OTLP exporter will wait for each batch export.
> 
>     Default: 10s
>     Env vars: OTEL_EXPORTER_OTLP_TIMEOUT OTEL_EXPORTER_OTLP_TRACES_TIMEOUT OTEL_EXPORTER_OTLP_METRICS_TIMEOUT OTEL_EXPORTER_OTLP_LOGS_TIMEOUT
> 

I'm omitting it here, but I did verify that setting `OTEL_EXPORTER_OTLP_TIMEOUT=3` resulting in that lag being three seconds.

Adding the one environment variable removes the lag completely:
```
$ time earthly-dev -P --no-cache +better-alpine
 <snip>
      +better-alpine | Mon Apr 29 16:46:54 UTC 2024
<snip>
      +better-alpine | Server:
      +better-alpine |  Engine:
      +better-alpine |   Version:          25.0.2
<snip>
      +better-alpine | Mon Apr 29 16:46:54 UTC 2024 going to stop dockerd
      +better-alpine | Mon Apr 29 16:46:56 UTC 2024 done stopping dockerd
      +better-alpine | time="2024-04-29T16:46:54.436348360Z" level=warning msg="WARNING: No swap limit support"
      +better-alpine | time="2024-04-29T16:46:54.436374634Z" level=info msg="Docker daemon" commit=fce6e0ca9bc000888de3daa157af14fa41fcd0ff containerd-snapshotter=false storage-driver=overlay2 version=25.0.2
      +better-alpine | time="2024-04-29T16:46:54.436463707Z" level=info msg="Daemon has completed initialization"
      +better-alpine | time="2024-04-29T16:46:54.508906959Z" level=info msg="API listen on /var/run/docker.sock"
      +better-alpine | time="2024-04-29T16:46:54.525733082Z" level=info msg="Processing signal 'terminated'"
      +better-alpine | time="2024-04-29T16:46:54.526368667Z" level=info msg="stopping event stream following graceful shutdown" error="<nil>" module=libcontainerd namespace=moby
      +better-alpine | time="2024-04-29T16:46:54.526889291Z" level=info msg="Daemon shutdown complete"
      +better-alpine | time="2024-04-29T16:46:54.526917872Z" level=info msg="stopping healthcheck following graceful shutdown" module=libcontainerd
      +better-alpine | time="2024-04-29T16:46:54.526946002Z" level=info msg="stopping event stream following graceful shutdown" error="context canceled" module=libcontainerd namespace=plugins.moby
      +better-alpine | ==== End dockerd logs ====
              output | --> exporting outputs
 <snip>
real	0m5.245s
user	0m0.282s
sys	0m0.146s
```

I would have created a PR, but wasn't sure if the obvious/easy solution would be the desired solution.  The simple solution is to simply add it to the wrapper script, but that would not allow power-users from using it (after they correctly configure it).  Or maybe that's good enough, but want it fed to the environment via the buildkit side (in earthly code).  Or it could be baked into earthly's opinionated d-in-d images, which would then allow power-users to use/configure their own d-in-d.

### What should have happened?

The end-user workaround is simply adding `ENV OTEL_TRACES_EXPORTER=none` into one's build.

### What earthly version?

This is more related to the version of docker in the d-in-d image being used.

**Comment by rrjjvv** [comment]
Also, I'm a creature of habit and I know my usage of plain `earthly/dind` and `earthly/dind:alpine` is considered deprecated.  The latter still receives updates, but it doesn't look like the former does.

If one is doing "the right thing" and using the fully tagged images posted on dockerhub, everyone is incurring that ten second penalty (as all of them have versions `25.x` or newer).
